File: py/portfolio.py
#!/usr/bin/env python
# encoding: utf-8
"""
@author: HuRuiFeng
@file: portfolio.py
@time: 2025/6/28 10:29
@project: tdd-project
@desc: 
"""

from money import Money


class Portfolio:
    def __init__(self):
        self.moneys = []

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank, currency):
        total = Money(0, currency)
        failures = ""
        for m in self.moneys:
            c, k = bank.convert(m, currency)
            if k is None:
                total += c
            else:
                failures += k if not failures else "," + k
        if not failures:
            return total

        raise Exception("Missing exchange rate(s):[" + failures + "]")

File: js/portfolio.js
/**
 * @author: HuRuiFeng
 * @file: portfolio.js
 * @time: 2025/6/28 10:03
 * @project: js
 * @desc:
 */

const Money = require('./money');

class Portfolio {
    constructor() {
        this.moneys = [];
    }

    add(...moneys) {
        this.moneys = this.moneys.concat(moneys);
    }

    evaluate(bank, currency) {
        // 初始化缺失的键的数组
        let failures = [];
        let total = this.moneys.reduce((sum, money) => {
            try {
                let convertedMoney = bank.convert(money, currency);
                return sum + convertedMoney.amount;
            } catch (error) {
                failures.push(error.message);
                return sum;
            }
        }, 0);
        if (!failures.length) {
            return new Money(total, currency);
        }
        throw new Error("Missing exchange rate(s):[" + failures.join() + "]");
    }
}

module.exports = Portfolio
